homeassistant/components/enocean/__init__.py

```diff
@@ -1,93 +1,66 @@
     """Set up the EnOcean component."""
-    serial_dev = config[DOMAIN].get(CONF_DEVICE)
-    dongle = EnOceanDongle(hass, serial_dev)
-    hass.data[DATA_ENOCEAN] = dongle
-
-    return True
-
-
-class EnOceanDongle:
-    """Representation of an EnOcean dongle."""
-
-    def __init__(self, hass, ser):
-        """Initialize the EnOcean dongle."""
-
-        self.__communicator = SerialCommunicator(port=ser, callback=self.callback)
-        self.__communicator.start()
-        self.hass = hass
-        self.hass.helpers.dispatcher.dispatcher_connect(
-            SIGNAL_SEND_MESSAGE, self._send_message_callback
+    # support for text-based configuration (legacy)
+    if DOMAIN not in config:
+        return True
+
+    configured_dongle = {
+        entry.data[CONF_DEVICE] for entry in hass.config_entries.async_entries(DOMAIN)
+    }
+    if configured_dongle:
+        # We can only have one dongle. If there is already one in the config,
+        # there is no need to import the yaml based config.
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config[DOMAIN]
         )
+    )
 
-    def _send_message_callback(self, command):
-        """Send a command through the EnOcean dongle."""
-        self.__communicator.send(command)
-
-    def callback(self, packet):
-        """Handle EnOcean device's callback.
-
-        This is the callback function called by python-enocan whenever there
-        is an incoming packet.
-        """
-
-        if isinstance(packet, RadioPacket):
-            _LOGGER.debug("Received radio packet: %s", packet)
-            self.hass.helpers.dispatcher.dispatcher_send(SIGNAL_RECEIVE_MESSAGE, packet)
-
+    return True
 
-class EnOceanDevice(Entity):
-    """Parent class for all devices associated with the EnOcean component."""
 
-    def __init__(self, dev_id, dev_name="EnOcean device"):
-        """Initialize the device."""
-        self.dev_id = dev_id
-        self.dev_name = dev_name
+async def async_setup_entry(
+    hass: core.HomeAssistant, config_entry: config_entries.ConfigEntry
+):
+    """Set up an EnOcean dongle for the given entry.
 
-    async def async_added_to_hass(self):
-        """Register callbacks."""
-        self.async_on_remove(
-            self.hass.helpers.dispatcher.async_dispatcher_connect(
-                SIGNAL_RECEIVE_MESSAGE, self._message_received_callback
-            )
-        )
+    This will read a flow-based configuration, but only if
+    a configuration was not found in the text-based configuration.
+    """
+    if DATA_ENOCEAN in hass.data and hass.data[DATA_ENOCEAN] is EnOceanDongle:
+        LOGGER.warning("Dongle already configured, skipping configuration")
+        return True
 
-    def _message_received_callback(self, packet):
-        """Handle incoming packets."""
+    if DATA_ENOCEAN not in hass.data:
+        hass.data[DATA_ENOCEAN] = {}
```

Comment:
```suggestion
    enocean_data = hass.data.setdefault(DATA_ENOCEAN, {})
```

Reply:
In ended up having to create a variable to store the default value, as setdefault was returning None when receiving the emty dict literal directly.
This was only happening when creating the integration though, not when starting hass with an already configured integration.
I am still investigating to understand why.

Reply:
Suggestion applied